In [ ]:
#Import necessory libraries
import pandas as pd
import re

import nltk
import string
from nltk.tokenize import word_tokenize, sent_tokenize

import warnings
warnings.filterwarnings('ignore')
%matplotlib inline

## Load the Data

In [ ]:
#Let's load the data 
df = pd.read_csv(r"C:\Users\ABHINANDAN\Desktop\Blackcoffer_article.csv")
df.head(10)

Unnamed: 0  URL_ID                                                URL  \
0           0       1  https://insights.blackcoffer.com/how-is-login-...   
1           1       2  https://insights.blackcoffer.com/how-does-ai-h...   
2           2       3  https://insights.blackcoffer.com/ai-and-its-im...   
3           3       4  https://insights.blackcoffer.com/how-do-deep-l...   
4           4       5  https://insights.blackcoffer.com/how-artificia...   
5           5       6  https://insights.blackcoffer.com/how-are-genet...   
6           6       7  https://insights.blackcoffer.com/how-is-ai-use...   
7           7       8  https://insights.blackcoffer.com/benefits-of-b...   
8           8       9  https://insights.blackcoffer.com/how-big-data-...   
9           9      10  https://insights.blackcoffer.com/how-will-ai-m...   

                                       Article_Title  \
0  How is Login Logout Time Tracking for Employee...   
1  How does AI help to monitor Retail Shelf watches?   
2          AI and its impact on the Fashion Industry   
3  How do deep learning models predict old and ne...   
4  How artificial intelligence can boost your pro...   
5  How are genetic sequencing maps affected by de...   
6        How is AI used to solve traffic management?   
7           Benefits of Big Data in Different fields   
8   How Big Data Will Impact the Future of Business?   
9     How Will AI Make Decisions In Tomorrow’s Wars?   

                                        Article_Text  
0  When people hear AI they often think about sen...  
1  With increasing computing power and more data,...  
2  If you were a fan of the 90’s film Clueless ba...  
3  Understanding exactly how data is ingested, an...  
4  From the stone age to the modern world, from h...  
5  Artificial intelligence (AI) is the developmen...  
6  Artificial intelligence (AI) is the most impor...  
7  “In God we trust, all others must bring data.”...  
8  Big data refers to large sets of unstructured,...  
9  Artificial intelligence (AI) is intelligence d...

In [ ]:
#lets check the shape and column names
print("Shape of the data :", df.shape)
print("Columns in data set :", df.columns)

Shape of the data : (170, 5)
Columns in data set : Index(['Unnamed: 0', 'URL_ID', 'URL', 'Article_Title', 'Article_Text'], dtype='object')


* Here we can see that our data set is having 170 rows and 5 columns, among these columns I will drop __Unnamed: 0__column as it just contains index.
* And I will combine __Article_Title__ and __Article_Text__ into single column as __Article__ and will drop both of these columns.

In [ ]:
#combining Article_Title and Article_Text into single column as Article
df['Article'] = df['Article_Title'].map(str)+' '+df['Article_Text']

#Now lets drop unwanted columns
df.drop(columns=["Unnamed: 0","Article_Title","Article_Text"], inplace=True)

df    #checking data after combining and dropping columns

URL_ID                                                URL  \
0         1  https://insights.blackcoffer.com/how-is-login-...   
1         2  https://insights.blackcoffer.com/how-does-ai-h...   
2         3  https://insights.blackcoffer.com/ai-and-its-im...   
3         4  https://insights.blackcoffer.com/how-do-deep-l...   
4         5  https://insights.blackcoffer.com/how-artificia...   
..      ...                                                ...   
165     167  https://insights.blackcoffer.com/role-big-data...   
166     168  https://insights.blackcoffer.com/sales-forecas...   
167     169  https://insights.blackcoffer.com/detect-data-e...   
168     170  https://insights.blackcoffer.com/data-exfiltra...   
169     171  https://insights.blackcoffer.com/impacts-of-co...   

                                               Article  
0    How is Login Logout Time Tracking for Employee...  
1    How does AI help to monitor Retail Shelf watch...  
2    AI and its impact on the Fashion Industry If y...  
3    How do deep learning models predict old and ne...  
4    How artificial intelligence can boost your pro...  
..                                                 ...  
165  Role of Big Data in Academia CAN ACADEMIA, RES...  
166  Statistical Methods for Sales Forecasting in R...  
167  Behavior Based Chi-Square model to Detect Data...  
168  What is Data Exfiltration? If we talk in terms...  
169  Impacts of COVID 19 on Food products Some vend...  

[170 rows x 3 columns]

In [ ]:
#lets check our first article
df.Article[0]

'How is Login Logout Time Tracking for Employees in Office done by AI? When people hear AI they often think about sentient robots and magic boxes. AI today is much more mundane and simple—but that doesn’t mean it’s not powerful. Another misconception is that high-profile research projects can be applied directly to any business situation. AI done right can create an extreme return on investments (ROIs)—for instance through automation or precise prediction. But it does take thought, time, and proper implementation. We have seen that success and value generated by AI projects are increased when there is a grounded understanding and expectation of what the technology can deliver from the C-suite down. “Artificial Intelligence (AI) is a science and a set of computational technologies that are inspired by—but typically operate quite differently from—the ways people use their nervous systems and bodies to sense, learn, reason and take action.”3 Lately there has been a big rise in the day-to-

## Data Cleaning

In [ ]:
#lets define a function to clean our text.
def clean(text):
    
    #removing editor's information
    x = re.search("Blackcoffer Insights", text)
    if x != None:
        a=x.span()
        text = text[:a[0]-1]
        
    #removing Share: from end
    text = re.sub("Share:",'',text)
    
    #removing urls
    text = re.sub("^http\://[a-zA-Z0-9-\.]+\.[a-zA-Z]{2,3}(/\S*)?$", '',text)
    #remove white space
    text = text.strip()
    
    return(text)


In [ ]:
#lets clean the text
df['Article'] = df['Article'].apply(lambda x : clean(x))

In [ ]:
#lets check our first article
df.Article[0]

'How is Login Logout Time Tracking for Employees in Office done by AI? When people hear AI they often think about sentient robots and magic boxes. AI today is much more mundane and simple—but that doesn’t mean it’s not powerful. Another misconception is that high-profile research projects can be applied directly to any business situation. AI done right can create an extreme return on investments (ROIs)—for instance through automation or precise prediction. But it does take thought, time, and proper implementation. We have seen that success and value generated by AI projects are increased when there is a grounded understanding and expectation of what the technology can deliver from the C-suite down. “Artificial Intelligence (AI) is a science and a set of computational technologies that are inspired by—but typically operate quite differently from—the ways people use their nervous systems and bodies to sense, learn, reason and take action.”3 Lately there has been a big rise in the day-to-

### Before removing stopwords from the text we will find the count for Personal Pronouns in the text and create a column for it.

In [ ]:
def pronoun_counter(text):
    words=word_tokenize(text)
    count=0
    for i in words:
        if i in ["I","we","my","ours","us"]:
            count+=1
    return(count)

In [ ]:
#lets clean the text
df['PRONOUN COUNT'] = df['Article'].apply(lambda x : pronoun_counter(x))

In [ ]:
df

URL_ID                                                URL  \
0         1  https://insights.blackcoffer.com/how-is-login-...   
1         2  https://insights.blackcoffer.com/how-does-ai-h...   
2         3  https://insights.blackcoffer.com/ai-and-its-im...   
3         4  https://insights.blackcoffer.com/how-do-deep-l...   
4         5  https://insights.blackcoffer.com/how-artificia...   
..      ...                                                ...   
165     167  https://insights.blackcoffer.com/role-big-data...   
166     168  https://insights.blackcoffer.com/sales-forecas...   
167     169  https://insights.blackcoffer.com/detect-data-e...   
168     170  https://insights.blackcoffer.com/data-exfiltra...   
169     171  https://insights.blackcoffer.com/impacts-of-co...   

                                               Article  PRONOUN COUNT  
0    How is Login Logout Time Tracking for Employee...              3  
1    How does AI help to monitor Retail Shelf watch...              2  
2    AI and its impact on the Fashion Industry If y...              7  
3    How do deep learning models predict old and ne...              1  
4    How artificial intelligence can boost your pro...             24  
..                                                 ...            ...  
165  Role of Big Data in Academia CAN ACADEMIA, RES...             14  
166  Statistical Methods for Sales Forecasting in R...              0  
167  Behavior Based Chi-Square model to Detect Data...              6  
168  What is Data Exfiltration? If we talk in terms...             11  
169  Impacts of COVID 19 on Food products Some vend...              1  

[170 rows x 4 columns]

## Removing stop_words from the text

### As assisted in problem statement, I will load file from given link contaning list for stopwords

In [ ]:
#loading the file
file = open("StopWords_GenericLong.txt",'r')
stop_words = file.read().split('\n')

#removing stopwords from text
df['Article'] = df['Article'].apply(lambda x : ' '.join([word for word in x.lower().split() if word not in (stop_words)]))

In [ ]:
df.Article[0]

'login logout time tracking employees office ai? people hear ai sentient robots magic boxes. ai today mundane simple—but doesn’t it’s powerful. misconception high-profile research projects applied directly business situation. ai create extreme return investments (rois)—for instance automation precise prediction. thought, time, proper implementation. success generated ai projects increased grounded understanding expectation technology deliver c-suite down. “artificial intelligence (ai) science set computational technologies inspired by—but typically operate differently from—the ways people nervous systems bodies sense, learn, reason action.”3 big rise day-to-day machines powered ai. machines wired cross-disciplinary approaches based mathematics, computer science, statistics, psychology, more.4 virtual assistants common, web shops predict purchases, companies make chatbots customer service companies algorithms detect fraud. ai deep learning technology employed office entry systems bring 

#### Great! now we are having our text without stopwords. Now we are good to go to do remaining analysis.

We will go for calculation of positive and negative score. For that we will create lists of positive and negative words, for this task I have used word dictionary that is mentioned in our problem statement.

In [ ]:
#getting Master Dictionary
Word_dict = pd.read_csv(r"Loughran-McDonald_MasterDictionary_1993-2021.csv")
Word_dict.head(5)

Word  Seq_num  Word Count  Word Proportion  Average Proportion  \
0   AARDVARK        1         354     1.550080e-08        1.422600e-08   
1  AARDVARKS        2           3     1.313627e-10        8.653817e-12   
2      ABACI        3           9     3.940882e-10        1.169679e-10   
3      ABACK        4          29     1.269840e-09        6.654735e-10   
4     ABACUS        5        8570     3.752595e-07        3.809464e-07   

        Std Dev  Doc Count  Negative  Positive  Uncertainty  Litigious  \
0  3.815486e-06         99         0         0            0          0   
1  9.241714e-09          1         0         0            0          0   
2  5.290465e-08          7         0         0            0          0   
3  1.595100e-07         28         0         0            0          0   
4  3.529356e-05       1108         0         0            0          0   

   Strong_Modal  Weak_Modal  Constraining  Syllables     Source  
0             0           0             0          2  12of12inf  
1             0           0             0          2  12of12inf  
2             0           0             0          3  12of12inf  
3             0           0             0          2  12of12inf  
4             0           0             0          3  12of12inf

In [ ]:
Word_dict.shape

(86531, 16)

In [ ]:
#Lets create lists for positive and negative words
positive=[]
negative = []
for idx in range(Word_dict.shape[0]):
    if Word_dict.Positive[idx]!=0:
        positive.append(Word_dict.Word[idx].lower())
    if Word_dict.Negative[idx]!=0:
        negative.append(Word_dict.Word[idx].lower())

In [ ]:
#lets check the content of our lists
print("Number of words in positive list: ", len(positive))
print("\nNumber of words in negative list: ", len(negative))
print("\nFirst 10 positive words in the positive list: ", positive[0:10])
print("\nFirst 10 negative words in the negative list: ", negative[0:10])

Number of words in positive list:  354

Number of words in negative list:  2355

First 10 positive words in the positive list:  ['able', 'abundance', 'abundant', 'acclaimed', 'accomplish', 'accomplished', 'accomplishes', 'accomplishing', 'accomplishment', 'accomplishments']

First 10 negative words in the negative list:  ['abandon', 'abandoned', 'abandoning', 'abandonment', 'abandonments', 'abandons', 'abdicated', 'abdicates', 'abdicating', 'abdication']


Great! now we are having two lists for postive and negative words.

In [ ]:
#lets create a function to calculate positive score for text

#first we will lower all text
df["Article"]=df["Article"].apply(lambda x: x.lower())

#defining a function to calculate positive score
def pos_score(text):
    count=0
    word_token = word_tokenize(text)
    for i in word_token:
        if i in positive:
            count+=1
    return count

#defining a function to calculate negative score
def neg_score(text):
    count=0
    word_token = word_tokenize(text)
    for i in word_token:
        if i in negative:
            count-=1
    return count*(-1)

In [ ]:
df["POSITIVE SCORE"]=df["Article"].apply(lambda x: pos_score(x))
df["NEGATIVE SCORE"]=df["Article"].apply(lambda x: neg_score(x))

In [ ]:
df.head(5)

URL_ID                                                URL  \
0       1  https://insights.blackcoffer.com/how-is-login-...   
1       2  https://insights.blackcoffer.com/how-does-ai-h...   
2       3  https://insights.blackcoffer.com/ai-and-its-im...   
3       4  https://insights.blackcoffer.com/how-do-deep-l...   
4       5  https://insights.blackcoffer.com/how-artificia...   

                                             Article  PRONOUN COUNT  \
0  login logout time tracking employees office ai...              3   
1  ai monitor retail shelf watches? increasing co...              2   
2  ai impact fashion industry fan 90’s film cluel...              7   
3  deep learning models predict drugs successfull...              1   
4  artificial intelligence boost productivity lev...             24   

   POSITIVE SCORE  NEGATIVE SCORE  
0               2               5  
1               8               6  
2              27              23  
3               5               1  
4              20              17

## Analysis of Readability

To analyse readability we will find Average Sentence Length, Percentage of Complex words and fog index

In [ ]:
#Average sentence length
def text_cleaner(text):
    #removeing punctuations from text
    txt = re.sub(r"[^\w\s]", "", text)
    #remove html markup
    txt = re.sub("(<.*?>)", "", txt)
    #remove non-ascii and digits
    txt = re.sub("(\\W)", " ", txt)
    txt = re.sub("(\\d)", "", txt)
    return txt

def avg_sent_len(text):
    sentence = sent_tokenize(text)
    txt = text_cleaner(text)
    words = word_tokenize(txt)
    
    return round((len(words)/len(sentence)),3)

def Word_count(text):
    #lets clean the text first
    txt = text_cleaner(text)
    words = word_tokenize(txt)
    return len(words)

#funftion to calculate syllable for a word
def syllable_count(word):
    count = 0
    vowels = ['a','e','i','o','u']
    if word[0] in vowels:
        count += 1
    for index in range(1, len(word)):
        if word[index] in vowels and word[index - 1] not in vowels:
            count += 1
            if word.endswith("e") or word.endswith('ed') or word.endswith('es'):
                count -= 1
    if count == 0:
        count += 1
    return count

#function to calculate number of complex words per text
def Count_complex_words(text):
    txt = text_cleaner(text) #text cleaning
    words = word_tokenize(txt)
    count = 0
    for i in words:
        if syllable_count(i)>2:
            count+=1
    return count

#function to calculate syllable per word
def Sl_per_word(text):
    txt = text_cleaner(text) #text cleaning
    words = word_tokenize(txt)
    sl_count=0
    #loop to count all syllables from text
    for i in words:
        syl =syllable_count(i)
        sl_count+=syl
    return round((sl_count/len(words)),3)

#function to calculate average word length
def Avg_word_len(text):
    txt = text_cleaner(text) #text cleaning
    words = word_tokenize(txt)
    word_len = 0
    for wrd in words:
        word_len+=len(wrd)
    return round((word_len/len(words)),3)


#function to calculate Average Number of Words Per Sentence
def words_per_sentence(text):
    sentence = sent_tokenize(text)
    txt = text_cleaner(text)
    words = word_tokenize(txt)
    return round((len(words)/len(sentence)),3)

In [ ]:
#Average sentence length
df["AVERAGE SENTENCE LENGTH"]=df['Article'].apply(lambda x: avg_sent_len(x))

#calculating number complex words
df["COMPLEX WORD COUNT"]=df['Article'].apply(lambda x: Count_complex_words(x))

#word count
df["WORD COUNT"]=df['Article'].apply(lambda x: Word_count(x))

#Count of sellables per word
df["SYLLABLE PER WORD"] = df['Article'].apply(lambda x: Sl_per_word(x))

#Percentage of complex words
df["PERCENTAGE OF COMPLEX WORDs"] = round(df["COMPLEX WORD COUNT"]/df["WORD COUNT"],3)

#fog index
df["FOG INDEX"] = round(0.4*(df['AVERAGE SENTENCE LENGTH']+df["PERCENTAGE OF COMPLEX WORDs"]),3)

#Average number of words per sentence
df["AVERAGE NUMBER OF WORDS PER SENTENCE"] = df['Article'].apply(lambda x: words_per_sentence(x))

#Average Word Length
df["AVERAGE WORD LENGTH"] = df['Article'].apply(lambda x: Avg_word_len(x))

## Calculate Polarity and Subjectivity scores

As the formulae for calculation polarity and subjectivity of the text are given, we will use those and calculate the same.

In [ ]:
#Polarity calculation
df["POLARITY SCORE"] = (df["POSITIVE SCORE"]-df["NEGATIVE SCORE"])/((df["POSITIVE SCORE"]+df["NEGATIVE SCORE"])+0.000001)

#Subjectivity calculation
df["SUBJECTIVITY SCORE"] = (df["POSITIVE SCORE"]+df["NEGATIVE SCORE"])/(df['WORD COUNT']+0.000001)

### Great! we have done all required analysis task. Let's have a look at our final results

In [ ]:
#
df

URL_ID                                                URL  \
0         1  https://insights.blackcoffer.com/how-is-login-...   
1         2  https://insights.blackcoffer.com/how-does-ai-h...   
2         3  https://insights.blackcoffer.com/ai-and-its-im...   
3         4  https://insights.blackcoffer.com/how-do-deep-l...   
4         5  https://insights.blackcoffer.com/how-artificia...   
..      ...                                                ...   
165     167  https://insights.blackcoffer.com/role-big-data...   
166     168  https://insights.blackcoffer.com/sales-forecas...   
167     169  https://insights.blackcoffer.com/detect-data-e...   
168     170  https://insights.blackcoffer.com/data-exfiltra...   
169     171  https://insights.blackcoffer.com/impacts-of-co...   

                                               Article  PRONOUN COUNT  \
0    login logout time tracking employees office ai...              3   
1    ai monitor retail shelf watches? increasing co...              2   
2    ai impact fashion industry fan 90’s film cluel...              7   
3    deep learning models predict drugs successfull...              1   
4    artificial intelligence boost productivity lev...             24   
..                                                 ...            ...   
165  role big data academia academia, researchers, ...             14   
166  statistical methods sales forecasting retail i...              0   
167  behavior based chi-square model detect data-ex...              6   
168  data exfiltration? talk terms general life, ex...             11   
169  impacts covid 19 food products vendors (fruit ...              1   

     POSITIVE SCORE  NEGATIVE SCORE  AVERAGE SENTENCE LENGTH  \
0                 2               5                   15.333   
1                 8               6                   12.071   
2                27              23                   12.750   
3                 5               1                   15.133   
4                20              17                    9.209   
..              ...             ...                      ...   
165              16              39                   10.620   
166              20              15                   14.737   
167               4              49                   10.623   
168               4              10                   10.214   
169               1               9                   14.885   

     COMPLEX WORD COUNT  WORD COUNT  SYLLABLE PER WORD  \
0                    69         368              1.728   
1                    75         338              1.822   
2                   225         969              1.835   
3                    50         227              1.872   
4                   146         617              1.823   
..                  ...         ...                ...   
165                 166         754              1.891   
166                 158         560              2.016   
167                 173         563              2.094   
168                  76         286              1.955   
169                  93         387              1.961   

     PERCENTAGE OF COMPLEX WORDs  FOG INDEX  \
0                          0.188      6.208   
1                          0.222      4.917   
2                          0.232      5.193   
3                          0.220      6.141   
4                          0.237      3.778   
..                           ...        ...   
165                        0.220      4.336   
166                        0.282      6.008   
167                        0.307      4.372   
168                        0.266      4.192   
169                        0.240      6.050   

     AVERAGE NUMBER OF WORDS PER SENTENCE  AVERAGE WORD LENGTH  \
0                                  15.333                6.829   
1                                  12.071                7.074   
2                                  12.750                7.221   
3                                  15.13

In [ ]:
#as article text is not required so I am dropping that column
df.drop(columns = 'Article', inplace=True)

In [ ]:
#saving the data into csv file
df.to_csv(r"C:\Users\ABHINANDAN\Desktop\Output Data Structure.xlsx")